xdsl/transforms/canonicalization_patterns/riscv.py

```diff
@@ -376,6 +376,53 @@ def match_and_rewrite(self, op: riscv.AddOp, rewriter: PatternRewriter) -> None:
             )
 
 
+class FuseMultiplyAddD(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: riscv.FAddDOp, rewriter: PatternRewriter) -> None:
+        """
+        Converts `c = a * b` and `d = e + c` to `d = e + a * b`.
+        `c` should not be used anywhere else and both operations must have the
+        `contract` fastmath flag set.
+        """
+
+        if (
+            op.fastmath is None
+            or llvm.FastMathFlag.ALLOW_CONTRACT not in op.fastmath.flags
+        ):
+            return
+
+        addend = mulop = None
+        if isinstance(op.rs1.type, riscv.RISCVRegisterType) and isinstance(
+            op.rs2.owner, riscv.FMulDOp
+        ):
+            addend = op.rs1
+            mulop = op.rs2.owner
+        elif isinstance(op.rs2.type, riscv.RISCVRegisterType) and isinstance(
+            op.rs1.owner, riscv.FMulDOp
+        ):
+            addend = op.rs2
+            mulop = op.rs1.owner
+        else:
+            return
+
+        if (
+            len(mulop.rd.uses) == 1
+            and (
+                mulop.fastmath is not None
+                and llvm.FastMathFlag.ALLOW_CONTRACT in mulop.fastmath.flags
+            )
+            and isinstance(mulop.rd.type, riscv.RISCVRegisterType)
+        ):
+            rewriter.replace_matched_op(
+                riscv.FMAddDOp(
+                    mulop.rs1,
+                    mulop.rs2,
+                    addend,
+                    comment=op.comment,
```

Comment:
```suggestion
                riscv.FMAddDOp(
                    mulop.rs1,
                    mulop.rs2,
                    addend,
                    rd=op.rd.type,
                    comment=op.comment,
```

Reply:
This might need another cast/isinstance check. It would be good to have tests for this also.

Reply:
Ah just saw this, I was working on the tests for the commutative cases.
Is this to transfer the register if we use this after register allocation?

Reply:
Yeah, there's a possibility that that happens, and then there's the danger of messing with the existing register allocation of the result value

Reply:
yeah, I was just planning the use of this prior to regalloc, but you're right, an oversight.